### DISCUSSION TITLE
non-cacheable fields

### DISCUSSION THREAD

**Discussion by deepsystm** [ElMassimo/oj_serializers#3] [Q&A] [answered]
Hi ElMassimo, thanks for your work, amazing and fast lib!

I have a few questions:

1. How do I make fields in the serializer that are always obtained by querying a value rather than a cached version of the value?

```ruby
class MixSerializer < Oj::Serializer
  has_one     :user, serializer: UserListItemSerializer

  attributes  :id, 
              :field_that_never_been_cached,
              :title


  attribute \
  def field_that_never_been_cached
    get_updated_data
  end

end
```

2. How to pass additional data to the serializer besides the serialization object itself? For example current_user object from Devise

```ruby
render json: AlbumSerializer.one(album, current_user: current_user) if signed_in?
```

Thanks!

**Comment by ElMassimo** [comment, marked as answer]
Hi Vladimir! @deepsystm

---

> How do I make fields in the serializer that are always obtained by querying a value rather than a cached version of the value?

By default, serializers do not cache any value, so every time the `MixSerializer` is used, it would call `get_updated_data` again.

If you share an example that's not working as expected in a public repo, I can take a look.

---

> How to pass additional data to the serializer besides the serialization object itself?

Any additional options, such as `current_user: current_user` are available in an `options` hash:

```ruby
class AlbumSerializer < Oj::Serializer
  def current_user
    options[:current_user]
  end
end
```

If it's something that you plan on using often, you can define helpers in a `BaseSerializer`.

Also, if the additional data is not optional, you might want to use `fetch` instead:

```ruby
  def current_user
    options.fetch(:current_user)
  end
```

That way it will fail early if the serializer is used incorrectly.

**Reply by deepsystm** [reply]
Thanks a lot! It's quite well explained! No more questions :1st_place_medal:

**Reply by deepsystm** [reply]
Hello, everyone!
I continue to use the library for development, it has shown great results in speed and ease of operation.

There is a question to continue the topic of passing parameters to the serializer:
**How can I pass parameters when calling has_one and has_many in the serializer itself?**

```
class MySerializer < Oj::Serializer

  has_one     :user, serializer: UserSerializer, my_parameter: options.fetch(:my_parameter, nil)

end


class UserSerializer < Oj::Serializer

  def my_parameter
    options.fetch(:my_parameter, nil)
  end

end

MySerializer.one(object, my_parameter: MY_VALUE)
```

when I try to do it, ruby says:

```error at line 2: undefined local variable or method `options' for MySerializer:Class```

I solve this this using something like:
```
class MySerializer < Oj::Serializer

  def user
       UserSerializer.one(object.user, my_parameter: MY_VALUE)
  end

end
```

But docs says:
**The associations DSL is more concise and achieves better performance, so prefer to use it instead of manually definining attributes**

**Reply by ElMassimo** [reply]
Hi Vladimir, glad it's been useful.

If you need to pass options, defining a method manually is the correct approach 👍 

---

To clarify on that "performance note", the difference is that when using `has_one` and `has_many` it will reuse the same string writer, while when calling `one` or `many` manually it will create a new string writer. Even then, writing the contents of a string writer is very efficient, so the performance difference should be negligible in most use cases.

**Reply by deepsystm** [reply]
thanks! :+1:
